商品出品機能

Kei-bigginer opened the pull request:
# what
ユーザーが商品情報を入力し、出品できる機能


# why
販売する商品を登録するため

### ①ログイン状態の場合は、商品出品ページへ遷移できる動画
https://gyazo.com/b6164d78ce1bfde090c2b392ed142e52
### ②価格が入力されると同時に、販売手数料と販売利益が表示される動画
https://gyazo.com/fb9af199a4ca3694ebf81e01a7213ef4
### ③必要な情報を適切に入力して「出品する」ボタンを押すと、商品情報がデータベースに保存される動画
https://gyazo.com/7dd1284ed3660c3eae5adc13c2d0ddf4
※一覧表示はまだなのでデータベースに反映されている動画添付しますが
画像の加工してないためactive_storage_variant_recordsには入っていません※
https://gyazo.com/29edb7ee074e218bca8bfa879dd98f0e


### ④入力に問題がある状態で「出品する」ボタンが押された場合、情報は保存されず、出品ページに戻りエラーメッセージが表示される動画
https://gyazo.com/fec342d1f4971571645072e2c5d5c344
### ⑤ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/7a02d28403c7e4b6a0fca59372ea44e2
### ⑥テスト結果の画像
https://gyazo.com/5771cca9ac72e6e4d9fb5a3275d1bd39

Kei-bigginer committed:
現時点でのitemモデルのバリデーション

Kei-bigginer committed:
Active Storageと画像加工ツールの導入は完了

Kei-bigginer committed:
Active Storageの導入

Kei-bigginer committed:
newのファイル少し書き直し

Kei-bigginer committed:
アクティブハッシュ（カテゴリ）済

Kei-bigginer committed:
アクティブハッシュおそらく済

Kei-bigginer committed:
アクティブストレージのバリデーションとエラー文作成済

Kei-bigginer committed:
JSによる利益と手数料計算記述済

Kei-bigginer committed:
モデルの単体テストコード済

Kei-bigginer committed:
不要な記述削除

OhyaYouki commented on app/controllers/application_controller.rb:
```diff
@@ -1,18 +1,22 @@
 class ApplicationController < ActionController::Base
   before_action :basic_auth
+  before_action :authenticate_user!, only: [:new, :create]
```
```ruby
before_action :authenticate_user!, only: [:new, :create]
```
上記はitemsコントローラーに記載しましょう。

OhyaYouki commented on app/javascript/item_price.js:
```diff
@@ -0,0 +1,17 @@
+window.addEventListener(`turbo:load`, () => {
```
商品の出品に失敗して出品ページに戻った際もjavascriptが動作するよう、renderメソッドに対応したイベントを追加しましょう。
以下カリキュラムのSTEP2、手順6をご参考ください。

実装に悩んだときのヒント
商品出品機能
https://master.tech-camp.in/v2/curriculums/8138

OhyaYouki commented on app/models/item.rb:
```diff
@@ -0,0 +1,60 @@
+    format: {with: /\A[0-9]+\z/, message:":半角数字で入力してください" }
+
+
+  validates :user, presence: true
```
```ruby
  validates :user, presence: true
```
上記は削除しましょう。

userに関するバリデーションはbelongs_toのアソシエーションが兼ねています。

OhyaYouki commented on app/models/item.rb:
```diff
@@ -0,0 +1,60 @@
+     greater_than_or_equal_to: 300 ,
+     less_than_or_equal_to: 9999999,
+    message: ":価格は、¥300~¥9,999,999の間で指定してください"	},
+    format: {with: /\A[0-9]+\z/, message:":半角数字で入力してください" }
```
```ruby
format: {with: /\A[0-9]+\z/, message:":半角数字で入力してください" }
```
上記は削除しましょう。

integer型のデータに対して正規表現を適応できません。

OhyaYouki commented on app/views/items/new.html.erb:
```diff
@@ -5,10 +5,10 @@
 
     <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
-    <%# render 'shared/error_messages', model: f.object %>
+      <%= render 'shared/error_messages', model: f.object %>
```
```
<%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
実装案内コメントアウトは削除しましょう。
他にも残っているコメントアウトがある場合削除をお願いいたします。

Kei-bigginer replied:
たくさんご指摘ありがとうございます。
修正いたしました。
再度確認お願いします！

OhyaYouki reviewed (commented):
レビュー依頼確認しました。

コメントいたしましたので、ご確認をお願いいたします。

コードの修正後、コミット・プッシュを行った上で再度レビュー依頼をお送りください。
よろしくお願いいたします。

Kei-bigginer committed:
修正済み

・itemsコントローラー
・renderメソッドに対応したイベントを追加
・validates :user, presence: true削除
・format: {with: /\A[0-9]+\z/, message:":半角数字で入力してください" }
削除

・実装案内コメントアウト削除

OhyaYouki reviewed (commented):
修正点を確認いたしました。

**LGTM**とさせていただきます！
実装お疲れ様でした！

こちらのブランチは忘れずにMergeいただき、ローカルのMainにPullした後に
次の実装のためのブランチを切って次の課題に取り組みましょう！

引き続き、開発頑張っていきましょう！応援しております。
